extensions/wikia/Follow/FollowHelper.class.php

```diff
@@ -129,24 +77,104 @@ public static function emailNotification( $childTitle, $list, $namespace, $user,
 				$message,
 				$action
 			);
-			$oTask->queue();
+
+			$task->queue();
 		}
+	}
 
-		wfProfileOut( __METHOD__ );
+	/**
+	 * Create and return arrays of at most MAX_WATCHERS_PER_JOB watchers
+	 *
+	 * @param WikiaSQL $watcherSQL
+	 *
+	 * @return array
+	 */
+	private static function getWatcherSets( WikiaSQL $watcherSQL ) {
+		$currentPage = 0;
+		$watcherCount = 0;
+
+		$watcherSets = $watcherSQL->runLoop(
+			wfGetDB( DB_SLAVE ),
+			function ( &$watcherSets, stdClass $row ) use ( &$currentPage, &$watcherCount ) {
+				$title = $row->wl_title;
+				$user = $row->wl_user;
+				$watcherSets[$currentPage][$title][] = $user;
+
+				$watcherCount++;
+				if ( ( $watcherCount % self::MAX_WATCHERS_PER_JOB ) == 0 ) {
+					$currentPage++;
+				}
+			}
+		);
+
+		return $watcherSets;
 	}
 
+	/**
+	 * Generate the WikiaSQL object needed to select all the watchers
+	 *
+	 * @param int $userId
+	 * @param string $namespace
+	 * @param array $list
+	 *
+	 * @return WikiaSQL
+	 */
+	private static function getWatcherSQL( $userId, $namespace, $list) {
+		$watcherSQL = ( new WikiaSQL() )
+			->SELECT( 'wl_user', 'wl_title' )
+			->FROM( 'watchlist' )
+			->WHERE( 'wl_user' )->NOT_EQUAL_TO( $userId )
+			->AND_( 'wl_namespace' )->EQUAL_TO( $namespace )
+			->AND_( 'wl_title' )->IN( $list )
+			->AND_( 'wl_notificationtimestamp' )->IS_NULL();
 
-	/*
-	 * blogListingBuildRelation - hook after save of blogListing create relations in table
+		$timeAgo = self::getTimeBoundary();
+		if ( !empty( $timeAgo ) ) {
+			$watcherSQL->OR_( 'wl_notificationtimestamp' )->LESS_THAN( $timeAgo );
```

Comment:
I'm little bit concerned about condition execution order.

Isn't it ```some condition AND some other condition AND ... AND last and condition OR some or condition```?
Finally we'd get only limited number of records limited by ```AND``` plus (because of ```OR```) all rows that  meets this condition (completely separately)